tensorflow-framework/src/main/java/org/tensorflow/framework/metrics/impl/MetricsHelper.java

```diff
@@ -212,7 +229,431 @@
+      Map<ConfusionMatrixEnum, Assign<T>> varInitializers,
+      Operand<T> labels,
+      Operand<T> predictions,
+      Operand<TFloat32> thresholds,
+      Integer topK,
+      Integer classIndex,
+      Operand<T> sampleWeight,
+      boolean multiLabel,
+      Operand<T> labelWeights) {
+    if (multiLabel && labelWeights != null)
+      throw new IllegalArgumentException(
+          "labelWeights for multilabel data should be handled outside of updateConfusionMatrixVariables when multiLabel is true.");
+
+    if (variablesToUpdate == null || variablesToUpdate.isEmpty()) {
+      return Collections.EMPTY_LIST;
+    }
+
+    Operand<T> tLabels = labels;
+    Operand<T> tPredictions = predictions;
+    Operand<T> tSampleWeight = sampleWeight;
+
+    // size of the 0th dimension of thresholds
+    // reshape to scalar for operations later.
+    Operand<TInt32> numThresholds =
+        tf.reshape(tf.shape.size(thresholds, tf.constant(0)), tf.constant(Shape.scalar()));
+
+    // true if we will process thresholds as one-dimensional (possibly because we flatten them)
+    Operand<TBool> oneThresh;
+    if (multiLabel) {
+      oneThresh = tf.math.equal(tf.constant(1), tf.rank(thresholds));
+    } else {
+      // TODO handle Ragged Tensors????
+      // [y_pred,
+      //    y_true], _ = ragged_assert_compatible_and_get_flat_values([y_pred, y_true],
+      //                                                   sampleWeights)
+      oneThresh = tf.constant(true);
+    }
+
+    List<Op> controlOps = new ArrayList<>();
+    Operand<TInt32> axes = allAxes(tf, tPredictions);
+    controlOps.add(
+        tf.withSubScope("updateConfusionMatrixVariables-1")
+            .assertThat(
+                tf.reduceAll(
+                    tf.math.greaterEqual(
+                        tPredictions, cast(tf, tf.constant(0), tPredictions.type())),
+                    axes),
+                Collections.singletonList(tf.constant("predictions must be >= 0"))));
+    controlOps.add(
+        tf.withSubScope("updateConfusionMatrixVariables-2")
+            .assertThat(
+                tf.reduceAll(
+                    tf.math.lessEqual(tPredictions, cast(tf, tf.constant(1), tPredictions.type())),
+                    axes),
+                Collections.singletonList(tf.constant("predictions must be <= 1"))));
+
+    LossTuple<T> result =
+        LossesHelper.squeezeOrExpandDimensions(tf, tLabels, tPredictions, tSampleWeight);
+    tPredictions = result.getTarget();
+    tLabels = result.getLabels();
+    tSampleWeight = result.getSampleWeights();
+
+    if (!tPredictions.shape().isCompatibleWith(tLabels.shape()))
+      throw new IllegalArgumentException(
+          String.format(
+              "Shapes %s and %s are incompatible)",
+              tPredictions.shape().toString(), tLabels.shape().toString()));
+
+    if (topK != null) {
+      tPredictions = filterTopK(tf, tPredictions, topK);
+    }
+
+    if (classIndex != null) {
```

Comment:
Shouldn't we also slice `tSampleWeight` in this case? (I do see that the Python code doesn't.)

Reply:
The non- topK predictions will be zeroed out, so it doesn't make a difference.

Reply:
I mean slice by `classIndex` in the `classIndex != null` case.

Reply:
`tSampleWeight` is broadcast to `tPredictions` later in the routine which takes care of the shape mismatch after the column slicing.

line: 534
`tSampleWeight = tf.broadcastTo(tSampleWeight, tf.shape(tPredictions));`

Reply:
Here's the case I'm worried about. Suppose that `tPredictions` and `tSampleWeight` both have shape (N, Cx), where Cx is some number of class dimensions. Suppose we use `classIndex` to slice `tPredictions` down to shape (N, Dx), where Dx is like Cx except with one of its dimensions sliced down to size 1. (Assume it wasn't already size 1.) But `tSampleWeight` still has size > 1 in that dimension, so the `broadcastTo` will fail.

Reply:
Sample weights will get squeezed if weights rank is 1 more than predictions rank and
will get expanded, if predictions  rank is 1 less than weights rank. Otherwise, it is left as is. I don't think this fully answers your question though. 

It seems that the result will take on the shape of the weights, with the last dimension
matching the last dim of the source, for example:

```
y_pred= np.array([[[0,1], [2,3]],[[4,5],[6,7]]])
print(y_pred.shape)
# (2, 2, 2)
weights = np.array([[[[[0.5],[0.3]], [[0.4],[0.5]]]]])
print(weights.shape)
# (1, 1, 2, 2, 1)
x = weights * y_pred;
print(x.shape)
# (1, 1, 2, 2, 2)
print(x)
```
The resulting shape is weights.shape.take(rank-1) + y_pred.shape.size(-1)
Maybe they intended this behavior as the resulting shape will be the same as if there was no slicing at all.

Reply:
Here's the case I'm worried about:
~~~Python
fp = tf.Variable([0.])
vars = { metrics_utils.ConfusionMatrix.FALSE_POSITIVES : fp }
y_true = [[1, 0, 0], [1, 0, 0], [0, 1, 0], [0, 1, 0]]
y_pred = [[.6, .3, .1], [.3, .6, .1], [.6, .3, .1], [.3, .6, .1]]
weight = [[.1, .2, .4], [.1, .2, .4], [.1, .2, .4], [.1, .2, .4]]
metrics_utils.update_confusion_matrix_variables(
    vars, y_true, y_pred, [.5]
    , sample_weight=weight
    , class_id=0
)
print(fp.numpy())
~~~
This results in the following error:
~~~
. . .
~/projects/python-tensorflow-sandbox/venv/lib/python3.7/site-packages/tensorflow/python/keras/utils/metrics_utils.py in update_confusion_matrix_variables(variables_to_update, y_true, y_pred, thresholds, top_k, class_id, sample_weight, multi_label, label_weights)
    406   if sample_weight is not None:
    407     sample_weight = weights_broadcast_ops.broadcast_weights(
--> 408         math_ops.cast(sample_weight, dtype=variable_dtype), y_pred)
    409     weights_tiled = array_ops.tile(
    410         array_ops.reshape(sample_weight, thresh_tiles), data_tiles)
. . .
ValueError: weights can not be broadcast to values. values.rank=1. weights.rank=2. values.shape=(4,). weights.shape=(4, 3).
~~~
The problem is that `y_true` and `y_pred` were sliced down by `class_id`, while `weight` was not. You can see this by commenting out either of these two lines:
~~~
    , sample_weight=weight
    , class_id=0
~~~
With either of those lines commented out, it runs fine.